### PR TITLE
fix: Improvement `conf["viur.paramFilterFunction"]`

### PR DESCRIPTION
**Pull request by skoegl**
Remove the static filter that removes any parameters passed, starting with "_", to a configurable filter function which provides more flexibility.

### CODE REVIEW

File: src/viur/core/config.py [hunk 1]
```diff
@@ -195,6 +195,9 @@ def __setitem__(self, key, value):
     # Upper limit of the amount of parameters we accept per request. Prevents Hash-Collision-Attacks
     "viur.maxPostParamsCount": 250,
 
+    # Function which decides if a request param should be used or filtered out
+    "viur.paramFilterFunction": lambda key, value: True,
```
Comment (ArneGudermann) [inline src/viur/core/config.py:199]:
```suggestion
    "viur.paramFilterFunction": None,
```

File: src/viur/core/request.py [hunk 2]
```diff
@@ -444,7 +445,7 @@ def _route(self, path: str) -> None:
                 # someone tries to exploit unicode normalisation bugs)
                 raise errors.BadRequest()
 
-            if key.startswith("_"):  # Ignore keys starting with _ (like VI's _unused_time_stamp)
+            if not param_filter(key, value):
```
Comment (ArneGudermann) [inline src/viur/core/request.py:448]:
```suggestion
            if param_filter and param_filter(key, value):
```
So we not call the lamda all the times

File: src/viur/core/request.py [hunk 3]
```diff
@@ -435,6 +435,7 @@ def _route(self, path: str) -> None:
                 f" of {conf['viur.maxPostParamsCount']} allowed arguments per request"
             )
 
+        param_filter = conf["viur.paramFilterFunction"]
```
Comment (ArneGudermann) [inline src/viur/core/request.py:438]:
```suggestion
        param_filter = conf["viur.paramFilterFunction"]
        if param_filter and not callable(param_filter):
           raise ValueError(f"""{param_filter=} is not callable""")
```

**Review by ArneGudermann — CHANGES_REQUESTED** [review]

File: src/viur/core/config.py (hunk 1 above)
Comment (sveneberth) [inline src/viur/core/config.py:199]: We should keep the old behavior.
```suggestion
    "viur.paramFilterFunction": lambda key, value: key.startswith("_"),
```

**Review by sveneberth — CHANGES_REQUESTED** [review]

File: src/viur/core/request.py [hunk 4]
```diff
@@ -435,6 +435,9 @@ def _route(self, path: str) -> None:
...
+        if param_filter and not callable(param_filter):
+           raise ValueError(f"""{param_filter=} is not callable""")
```
Comment (phorward) [inline src/viur/core/request.py:440]:
```suggestion
           raise ValueError(f"""{param_filter=} is not callable""")

```

**Review by phorward — APPROVED** [review]

File: src/viur/core/request.py [hunk 5]
```diff
@@ -435,6 +435,10 @@ def _route(self, path: str) -> None:
...
```
Comment (phorward) [inline src/viur/core/request.py:440]:
```suggestion
            raise ValueError(f"""{param_filter=} is not callable""")
```

File: src/viur/core/request.py [hunk 6]
```diff
@@ -435,6 +435,10 @@ def _route(self, path: str) -> None:
                 f" of {conf['viur.maxPostParamsCount']} allowed arguments per request"
             )
 
+        param_filter = conf["viur.paramFilterFunction"]
+        if param_filter and not callable(param_filter):
+            raise ValueError(f"""{param_filter=} is not callable""")
```
Comment (sveneberth) [inline src/viur/core/request.py:440]: I think this check could be moved to a better place where it is not called on every request. But okay for `main`. On `develop`, we can validate it in the `config` object.

**Review by sveneberth — APPROVED** [review]

File: src/viur/core/request.py (hunk 6 above)
Comment (phorward) [reply src/viur/core/request.py:440]: >  On `develop`, we can validate it in the `config` object.

Can you either tell me how, or make the merge?

Comment (sveneberth) [reply src/viur/core/request.py:440]: Maybe with an `@property` like https://github.com/viur-framework/viur-core/blob/80b93191df5d6d8c426ac5181f8d6696c666c168/src/viur/core/config.py#L717-L747

In the longer term, we could also look to validate the values in setattr against the type hints.

**Pull request merged**